File: src/main/kotlin/de/gieskerb/class_craft/data/PlayerData.java
package de.gieskerb.class_craft.data;

import de.gieskerb.class_craft.Main;
import de.gieskerb.class_craft.classes.BaseClass;
import de.gieskerb.class_craft.listener.HungerChangeListener;
import org.bukkit.Bukkit;
import org.bukkit.scheduler.BukkitTask;
import org.json.simple.JSONObject;

import java.util.ArrayList;
import java.util.Objects;

public class PlayerData {


    private final static ArrayList<PlayerData> allPlayerData = new ArrayList<>();
    private final String assignedPlayer;
    private BaseClass activeBaseClass;
    private HorseData horseData;
    private transient DistanceTracker distanceTracker;
    private transient BukkitTask distanceTrackerTask;

    public PlayerData(String playerName) {
        this.assignedPlayer = playerName;
        this.activeBaseClass = null;
        this.horseData = null;
        this.distanceTracker = null;
    }

    public PlayerData(JSONObject json) {
        this.assignedPlayer = json.get("player").toString();
        this.activeBaseClass = BaseClass.fromJSON((JSONObject) json.get("class"), Bukkit.getPlayer(assignedPlayer));
        this.horseData = HorseData.fromJSON((JSONObject) json.get("horse"));
    }

    public static void addPlayerData(PlayerData playerData) {
        allPlayerData.add(playerData);
    }

    public static PlayerData getPlayerData(String name) {
        if (name == null) return null;
        for (PlayerData playerData : allPlayerData) {
            if (name.equals(playerData.assignedPlayer)) return playerData;
        }
        return null;
    }

    public static void removePlayerData(String name) {
        if (name == null) return;
        allPlayerData.removeIf(playerData -> name.equals(playerData.assignedPlayer));
    }


    public JSONObject toJSON() {
        JSONObject json = new JSONObject();
        json.put("player", assignedPlayer);
        if (this.activeBaseClass != null) {
            json.put("class", activeBaseClass.toJSON());
        } else {
            json.put("class", new JSONObject());
        }
        if (this.horseData != null) {
            json.put("horse", horseData.toJSON());
        } else {
            json.put("horse", new JSONObject());
        }
        return json;
    }

    public DistanceTracker getDistanceTracker() {
        return this.distanceTracker;
    }

    public void setDistanceTracker(DistanceTracker distanceTracker) {
        if (distanceTrackerTask != null) {
            this.distanceTrackerTask.cancel();
        }
        if (this.distanceTracker != null) {
            this.distanceTracker.despawnHorse();
        }

        this.distanceTracker = distanceTracker;
        this.distanceTrackerTask = Bukkit.getScheduler().runTaskTimer(Main.getPlugin(), () -> {
            if (this.distanceTracker.trackDistance()) {
                this.distanceTrackerTask.cancel();
                this.distanceTracker = null;
                this.distanceTrackerTask = null;
            }
        }, 0L, 20L);
    }

    public void changeClass(BaseClass activeBaseClass) {
        if (activeBaseClass == null) {
            throw new NullPointerException("activeBaseClass is null -> use resetClass instead!");
        }
        if (this.activeBaseClass != null) {
            this.resetClass();
        }
        this.activeBaseClass = activeBaseClass;
        Objects.requireNonNull(Bukkit.getPlayer(assignedPlayer)).sendMessage(
                "§aYou selected the §6" + this.activeBaseClass.CLASS_NAME + " Class");
    }

    public void resetClass() {
        if (this.activeBaseClass != null) {
            this.activeBaseClass.removePermanentEffects();
        }
        HungerChangeListener.removePlayer(assignedPlayer);
        this.activeBaseClass = null;
        this.horseData = null;
    }

    public BaseClass getActiveClass() {
        return this.activeBaseClass;
    }

    public HorseData getHorseData() {
        return this.horseData;
    }

    public void setHorseData(HorseData horseData) {
        this.horseData = horseData;
    }

}

File: src/main/kotlin/de/gieskerb/class_craft/data/PlayerData.kt
package de.gieskerb.class_craft.data

import de.gieskerb.class_craft.Main
import de.gieskerb.class_craft.classes.BaseClass
import de.gieskerb.class_craft.listener.HungerChangeListener
import org.bukkit.Bukkit
import org.bukkit.scheduler.BukkitTask
import org.json.simple.JSONObject

class PlayerData {
    val assignedPlayer: String
    var activeClass: BaseClass?
        private set

    var horseData: HorseData?

    @Transient
    private var distanceTracker: DistanceTracker? = null

    @Transient
    private var distanceTrackerTask: BukkitTask? = null

    constructor(playerName: String) {
        this.assignedPlayer = playerName
        this.activeClass = null
        this.horseData = null
        this.distanceTracker = null
    }

    constructor(json: JSONObject) {
        this.assignedPlayer = json.get("player").toString()
        this.activeClass = Bukkit.getPlayer(assignedPlayer)?.let { BaseClass.fromJSON(json["class"] as JSONObject, it) }
        this.horseData = HorseData.fromJSON(json["horse"] as JSONObject)
    }

    fun toJSON(): JSONObject {
        val json = JSONObject()
        json["player"] = assignedPlayer
        if (this.activeClass != null) {
            json["class"] = activeClass!!.toJSON()
        } else {
            json["class"] = JSONObject()
        }
        if (this.horseData != null) {
            json["horse"] = horseData!!.toJSON()
        } else {
            json["horse"] = JSONObject()
        }
        return json
    }

    fun getDistanceTracker(): DistanceTracker? {
        return this.distanceTracker
    }

    fun setDistanceTracker(distanceTracker: DistanceTracker?) {
        if (distanceTrackerTask != null) {
            distanceTrackerTask!!.cancel()
        }
        if (this.distanceTracker != null) {
            this.distanceTracker!!.despawnHorse()
        }

        this.distanceTracker = distanceTracker
        this.distanceTrackerTask = Bukkit.getScheduler().runTaskTimer(
            Main.plugin, Runnable {
                if (this.distanceTracker!!.trackDistance()) {
                    distanceTrackerTask!!.cancel()
                    this.distanceTracker = null
                    this.distanceTrackerTask = null
                }
            }, 0L, 20L
        )
    }

    fun changeClass(activeBaseClass: BaseClass) {
        this.resetClass()
        this.activeClass = activeBaseClass
        Bukkit.getPlayer(assignedPlayer)?.sendMessage(
            "§aYou selected the §6" + activeClass!!.CLASS_NAME + " Class"
        )
    }

    fun resetClass() {
        activeClass?.removePermanentEffects()
        HungerChangeListener.removePlayer(assignedPlayer)
        this.activeClass = null
        this.horseData = null
    }

    companion object {
        val allPlayerData = ArrayList<PlayerData>()
        fun addPlayerData(playerData: PlayerData) {
            allPlayerData.add(playerData)
        }

        fun getPlayerData(name: String?): PlayerData? {
            if (name == null) return null
            for (playerData in allPlayerData) {
                if (name == playerData.assignedPlayer) return playerData
            }
            return null
        }

        fun removePlayerData(name: String?) {
            if (name == null) return
            allPlayerData.removeIf { playerData: PlayerData -> name == playerData.assignedPlayer }
        }
    }
}